controlplane/pkg/common/mechanism_selector.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2019 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"strconv"
+
+	"github.com/pkg/errors"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection/mechanisms/vxlan"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/networkservice"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/model"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/vni"
+)
+
+// MechanismSelector is for selecting and finding forwarding mechanisms. It is used in forwarder service.
+type MechanismSelector interface {
+	Select(*networkservice.NetworkServiceRequest, *model.Forwarder) (*connection.Mechanism, error)
+	Find(*model.Forwarder, string) *connection.Mechanism
+}
+
+// RemoteMechanismSelector for remote mechanisms
+type RemoteMechanismSelector struct {
+	vniAllocator vni.VniAllocator
+}
+
+// LocalMechanismSelector for remote mechanisms
+type LocalMechanismSelector struct{}
+
+func findMechanism(mechanismPreferences []*connection.Mechanism, mechanismType string) *connection.Mechanism {
+	for _, m := range mechanismPreferences {
+		if m.GetType() == mechanismType {
+			return m
+		}
+	}
+	return nil
+}
+
+// Select remote mechanisms of particular type from forwarder
+func (selector *RemoteMechanismSelector) Select(request *networkservice.NetworkServiceRequest, fwd *model.Forwarder) (*connection.Mechanism, error) {
+	for _, mechanism := range request.GetRequestMechanismPreferences() {
+		fwdMechanism := selector.Find(fwd, vxlan.MECHANISM)
```

Comment:
this can be put before the loop, because selector.Find() depends only on 'fwd' argument

Reply:
Well, this comes from old `controlplane/pkg/remote/forwarder_service.go`, and looks kinda strange. As far as I understand, this is because now only `vxlan.MECHANISM` is supported for the remote case. But I think we need to replace this constant with variable `mechanism.Type` here